src/annotator/frame-observer.js

```diff
@@ -101,44 +101,140 @@ export class FrameObserver {
 }
 
 /**
- * Resolves a Promise when the iframe's document is ready (loaded and parsed)
+ * Test if this is the empty document that a new iframe has before the URL
+ * specified by its `src` attribute loads.
  *
  * @param {HTMLIFrameElement} frame
- * @return {Promise<void>}
- * @throws {Error} if trying to access a document from a cross-origin iframe
  */
-export function onDocumentReady(frame) {
-  return new Promise(resolve => {
-    // @ts-expect-error
-    const frameDocument = frame.contentWindow.document;
-    const { readyState, location } = frameDocument;
-
-    // Web browsers initially load a blank document before the final document.
-    // This blank document is (1) accessible, (2) has an empty body and head,
-    // and (3) has a 'complete' readyState, on Chrome and Safari, and an
-    // 'uninitialized' readyState on Firefox. If a blank document is detected and
-    // there is a 'src' attribute, it is expected that the blank document will be
-    // replaced by the final document.
-    if (
-      location.href === 'about:blank' &&
-      frame.hasAttribute('src') &&
-      frame.src !== 'about:blank'
-    ) {
-      // Listening to `DOMContentLoaded` on the frame's document doesn't work because the
-      // document is replaced. On the other hand, listening the frame's `load`
-      // works because the frame element (as in HTMLIFrameElement) doesn't change.
-      frame.addEventListener('load', () => {
-        resolve();
-      });
+function hasBlankDocumentThatWillNavigate(frame) {
+  return (
+    frame.contentDocument?.location.href === 'about:blank' &&
+    // Do we expect the frame to navigate away from about:blank?
+    frame.hasAttribute('src') &&
+    frame.src !== 'about:blank'
+  );
+}
+
+/**
+ * Wrapper around {@link onDocumentReady} which returns a promise that resolves
+ * the first time that a document in `frame` becomes ready.
+ *
+ * See {@link onDocumentReady} for the definition of _ready_.
+ *
+ * @param {HTMLIFrameElement} frame
+ * @return {Promise<Document>}
+ */
+export function onNextDocumentReady(frame) {
+  return new Promise((resolve, reject) => {
+    const unsubscribe = onDocumentReady(frame, (err, doc) => {
```

Comment:
Had to go look at the implementation of `onDocumentReady` to grok how `unsubscribe` becomes a function reference "before anything else can happen here".